fs/shm/shmfs.c

```diff
@@ -0,0 +1,384 @@
+    {
+      return -EINVAL;
+    }
+
+  ret = inode_lock();
+  if (ret >= 0)
+    {
+      object = filep->f_inode->i_private;
+      if (!object)
+        {
+          filep->f_inode->i_private = alloc_shm_object(length);
+        }
+      else if (object->length != length)
+        {
+          /* This doesn't support resize */
+
+          ret = -EINVAL;
+        }
+
+      inode_unlock();
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: shmfs_unlink
+ ****************************************************************************/
+
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+static int shmfs_unlink(FAR struct inode *inode)
+{
+  if (inode->i_crefs <= 1)
+    {
+      delete_shm_object(inode->i_private);
+      inode->i_private = NULL;
+    }
+
+  return OK;
+}
+#endif
+
+/****************************************************************************
+ * Name: shmfs_map_object
+ ****************************************************************************/
+
+static int shmfs_map_object(FAR struct shmfs_object_s *object,
+                            FAR void **vaddr)
+{
+  int ret = OK;
+
+#ifdef CONFIG_BUILD_KERNEL
+  /* Map the physical pages of the shm object with MMU. */
+
+  FAR struct tcb_s        *tcb   = nxsched_self();
+  FAR struct task_group_s *group = tcb->group;
+  FAR uintptr_t *pages =
+    (FAR uintptr_t *)&object->paddr;
+  uintptr_t                mapaddr;
+  unsigned int             npages;
+
+  /* Find a free vaddr space that satisfies length */
+
+  mapaddr = (uintptr_t)vm_alloc_region(get_group_mm(group), 0,
+                                       object->length);
+  if (mapaddr == 0)
+    {
+      return -ENOMEM;
+    }
+
+  /* Convert the region size to pages */
+
+  npages = MM_NPAGES(object->length);
+
+  /* Map the memory to user virtual address space */
+
+  ret = up_shmat(pages, npages, mapaddr);
+  if (ret < 0)
+    {
+      vm_release_region(get_group_mm(group), (FAR void *)mapaddr,
+                        object->length);
+    }
+  else
+    {
+      *vaddr = (FAR void *)mapaddr;
+    }
+#else
+  /* Use the physical address directly */
+
+  *vaddr = object->paddr;
+#endif
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: shmfs_mmap
+ ****************************************************************************/
+
+static int shmfs_mmap(FAR struct file *filep,
+                      FAR struct mm_map_entry_s *entry)
+{
+  FAR struct shmfs_object_s *object;
+  int ret = -EINVAL;
+
+  DEBUGASSERT(filep->f_inode != NULL);
+
+  /* We don't support offset at the moment, just mapping the whole object */
+
+  if (entry->offset != 0)
+    {
+      return ret;
+    }
+
+  /* Keep the inode when mmapped, increase refcount */
+
+  ret = inode_addref(filep->f_inode);
+  if (ret >= 0)
+    {
+      object = (FAR struct shmfs_object_s *)filep->f_inode->i_private;
+
+      if (object)
```

Comment:
move the check before line 271 and return fail directly

Reply:
I think that the c_refs need to be bumped first before tinkering with filep->f_inode

Reply:
why? filep already hold one reference for you

Reply:
Is it the opened file (descriptor) that keeps the reference ? What happens now when user calls close(fd) ? Does the private data get also destroyed, even if someone has it mapped ?

Reply:
That's right! It may be bad application design, but it is possible to share file descriptors between threads in the same application. You can't rely on refcount on the open call only here. Of course the application wouldn't work, but it is better that mmap either succeeds or fails, and doesn't crash.

Reply:
reverted the earlier change, now object is back inside the refcount bump

Reply:
I think the following is perfectly legal code:
```

int fd = open("myfile", XX);
void *ptr = mmap(NULL, 1024, PROT_XX, MAP_XX, 0);
close(fd);

sprintf(ptr, "Foobar!\n",);
printf("%s" (const char *)ptr);

```
etc

So if the open file descriptor is the only one keeping the mapped data alive, the above code will fail.

Reply:
yes of course. This wasn't about not bumping the refcounter in mmap at all, it was just the place where you use the "object" ptr (inode->i_private). I think that is is supposed to be after refcount bump. otherwise crazy multithreaded application can close the fd in the middle of other thread doing mmap on it.

Reply:
Now I get it.